### PR TITLE
Initial library iteration

### PR DESCRIPTION
**Pull request by Meldiron**

### CODE REVIEW

File: docs/tutorials/add-logger-adapter.md [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        // TODO: Fill protected variables with keys using values from constructor parameters
+    }
+}
+```
+
+> If you copy this template, make sure to replace all placeholders wrapped like `[THIS]` and to implement everything marked as `TODO:`.
+
+When implementing new adapter, please make sure to follow these rules:
+
+- `getAdapterName()` needs to use same name as file name with first letter lowercased. For example, in `AppSignal.php`, we use `appSignal`
```
Comment (eldadfux) [inline docs/tutorials/add-logger-adapter.md:111]:
```suggestion
- `getName()` needs to use same name as file name with first letter lowercased. For example, in `AppSignal.php`, we use `appSignal`
```

File: src/Logger/Adapter/Raygun.php [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+     * @var string (required, can be found in Appsignal -> Project -> App Settings -> Push & deploy -> Push Key)
+     */
+    protected string $apiKey;
+
+    /**
+     * Return unique adapter name
+     *
+     * @return string
+     */
+    public static function getAdapterName(): string
```
Comment (eldadfux) [inline src/Logger/Adapter/Raygun.php:25]:
```suggestion
    public static function getName(): string
```

The scope already suggests this is the adapter name.

File: src/Logger/Adapter/Sentry.php [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+     * @var string (required, this part of Sentry DSN: 'https://blabla@blabla.ingest.sentry.io/{{THIS_PART}}')
+     */
+    protected string $projectId;
+
+    /**
+     * Return unique adapter name
+     *
+     * @return string
+     */
+    public static function getAdapterName(): string
```
Comment (eldadfux) [inline src/Logger/Adapter/Sentry.php:30]:
```suggestion
    public static function getName(): string
```

The scope already suggests this is the adapter name.

File: docs/tutorials/add-logger-adapter.md [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+class [ADAPTER_NAME] extends Adapter
+{
+    // TODO: Define protected variables with keys required for authentication with external Adapter API
+
+    /**
+     * Return unique adapter name
+     *
+     * @return string
+     */
+    public static function getAdapterName(): string
```
Comment (eldadfux) [inline docs/tutorials/add-logger-adapter.md:67]:
```suggestion
    public static function getName(): string
```

File: src/Logger/Adapter.php [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+use Exception;
+
+abstract class Adapter
+{
+    /**
+     * Get unique name of an adapter
+     *
+     * @return string
+     */
+    abstract public static function getAdapterName(): string;
```
Comment (eldadfux) [inline src/Logger/Adapter.php:14]:
```suggestion
    abstract public static function getName(): string;
```

The scope already suggests this is the adapter name.

File: src/Logger/Adapter/AppSignal.php [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+     * @var string (required, can be found in Appsignal -> Project -> App Settings -> Push & deploy -> Push Key)
+     */
+    protected string $apiKey;
+
+    /**
+     * Return unique adapter name
+     *
+     * @return string
+     */
+    public static function getAdapterName(): string
```
Comment (eldadfux) [inline src/Logger/Adapter/AppSignal.php:25]:
```suggestion
    public static function getName(): string
```

The scope already suggests this is the adapter name.

File: src/Logger/Adapter/AppSignal.php [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+        );
+
+        // apply those options
+        curl_setopt_array($ch, $optArray);
+
+        // execute request and get response
+        $result = curl_exec($ch);
+
+        // also get the error and response code
+        $errors = curl_error($ch);
```
Comment (eldadfux) [inline src/Logger/Adapter/AppSignal.php:119]: I think we should handle HTTP errors here, and throw an Exception if something went wrong. True to all adapters.

File: src/Logger/Adapter/Raygun.php [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+     */
+    public function __construct(string $configKey)
+    {
+        $this->apiKey = $configKey;
+    }
+
+    public function validateLog(Log $log): bool
+    {
+        // Support all log types (as tag)
+        switch ($log->getType()) {
```
Comment (eldadfux) [inline src/Logger/Adapter/Raygun.php:125]: If this is a repeating trend, I think we should have an array containing the allowed log types per adapter. Then we can implement `isLogSupported` method (or a better name) on the abstract adapter directly. This method could also be useful for dev using the lib, it can allow them to know what might be supported before getting errors.

File: src/Logger/Adapter/Raygun.php [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+                case Log::TYPE_ERROR:
+                case Log::TYPE_WARNING:
+                    break;
+                default:
+                    throw new Exception("Supported breadcrumb types for this adapter are: TYPE_INFO, TYPE_DEBUG, TYPE_VERBOSE, TYPE_ERROR, TYPE_WARNING");
+            }
+        }
+
+        // support all environment types (as tag)
+        switch ($log->getEnvironment()) {
```
Comment (eldadfux) [inline src/Logger/Adapter/Raygun.php:151]: We can have the same logic as suggested for log types. We have a similar approach for db features in `utopia/database` for checking what abilities each adapter supports.

**Review by eldadfux — CHANGES_REQUESTED** [review]

File: src/Logger/Adapter/AppSignal.php (hunk 6 above)
Comment (Meldiron) [reply src/Logger/Adapter/AppSignal.php:25]: Renamed all to `getName`.

File: src/Logger/Adapter.php (hunk 5 above)
Comment (Meldiron) [reply src/Logger/Adapter.php:14]: Renamed all to `getName`.

File: docs/tutorials/add-logger-adapter.md (hunk 4 above)
Comment (Meldiron) [reply docs/tutorials/add-logger-adapter.md:67]: Renamed all to `getName`.

File: src/Logger/Adapter/Sentry.php (hunk 3 above)
Comment (Meldiron) [reply src/Logger/Adapter/Sentry.php:30]: Renamed all to `getName`.

File: src/Logger/Adapter/Raygun.php (hunk 2 above)
Comment (Meldiron) [reply src/Logger/Adapter/Raygun.php:25]: Renamed all to `getName`.

File: docs/tutorials/add-logger-adapter.md (hunk 1 above)
Comment (Meldiron) [reply docs/tutorials/add-logger-adapter.md:111]: Renamed all to `getName`.

File: src/Logger/Adapter/Raygun.php (hunk 8 above)
Comment (Meldiron) [reply src/Logger/Adapter/Raygun.php:125]: Not sure if everything can be implemented like this. For example, we would need `getSupportedLogTypes()`, `getSupportedBreadcrumbTypes()` and `getSupportedEnvironmentTypes()`.

- Some providers might support `any` environment type, so the function would have to return some hack-y value to say "anything supported".
- Some providers might not support breadcrumbs, so it's silly to try to return an array of supported breadcrumb types. 
- Some providers might have more options than we support, for example required field for the device name. How does a developer achieve that? Well, he would need to use `tags` or `extra` array to provide that information. Then, adapter needs to nave this special check where he looks at `tags` array and check if `deviceName` is present.

File: src/Logger/Adapter/Raygun.php (hunk 9 above)
Comment (Meldiron) [reply src/Logger/Adapter/Raygun.php:151]: Same as above applies. Checks can get specific in different providers and function `validateLog` gives the most flexibility.

File: src/Logger/Adapter/AppSignal.php (hunk 7 above)
Comment (Meldiron) [reply src/Logger/Adapter/AppSignal.php:119]: Added better error handling and throw errors.

**Comment by Meldiron** [comment]
Prepared GH issue for a provider. Please review it before I create an issue for each provider. https://github.com/utopia-php/logger/issues/4

File: src/Logger/Adapter/Raygun.php (hunk 8 above)
Comment (christyjacob4) [reply src/Logger/Adapter/Raygun.php:130]: @eldadfux currently I don't think it would be possible to make this as part of the abstract class since not all providers may have concepts like `breadcrumbType`, `logType`, `environment` . So we decided to use a generic `validateLog()` method so each provider can have a different way to validate their log. 

As we have more providers, we can better decide which methods need to make way to the abstract class. Let us know what you think

File: src/Logger/Adapter/Raygun.php (hunk 9 above)
Comment (Meldiron) [reply src/Logger/Adapter/Raygun.php:157]: We could do it. Could look like
```
const supportedLogTypes = [
   TYPE_DEBUG,
   TYPE_ERROR,.  ....
]

getSupportForLogTypes (string $type) {
         return if supportedLogTypes contains $type
}
```

With type and environment, it might make sense. But what about `getSupportForBreadcrumbTypes()`? There might be adapter that does not support at all. It doesn't support breadcrumb, but it has to have a function that returns supported types? That's the dilemma we are facing - if we want functions that are not fully general.

Comment (eldadfux) [reply src/Logger/Adapter/Raygun.php:157]: `getSupportForBreadcrumbTypes()` might work better if we change it to `getSupportForBreadcrumb(): bool`

Comment (christyjacob4) [reply src/Logger/Adapter/Raygun.php:157]: @eldadfux are you sure `getSupportForBreadcrumb()` and other specific functions would be a good idea to have in the abstract class ? As we mentioned before, not all providers would have concepts like breadcrumbs, logTypes, environmentTypes and so on. 

In the future, if another provider introduces a new concept like XYZ , we will need to add the `getSupportForXYZ()` to the abstract class and all other existing classes.  Do you think this would be a good pattern being too specific ? or we can keep it pretty open like `validate()`

**Comment by kodumbeats** [comment]
Does this library need to support PHP7.4? Can we run with PHP8 or 8.1?

**Comment by Meldiron** [comment]
> Does this library need to support PHP7.4? Can we run with PHP8 or 8.1?

@eldadfux Your decision probably. I would guess code can work without problem on both. With v8, there might be features to make code cleaner.

**Comment by eldadfux** [comment]
@Meldiron We can stick to 8.0 and 8.1.

File: src/Logger/Adapter/Raygun.php (hunk 9 above)
Comment (Meldiron) [reply src/Logger/Adapter/Raygun.php]: Implemented `getSupportedTypes`, `getSupportedEnvironments`, `getSupportedBreadcrumbTypes` in https://github.com/utopia-php/logger/commit/74140bf8bf9cb8aeb19f7505506e9cd9b71b725b

File: src/Logger/Adapter/Raygun.php (hunk 8 above)
Comment (Meldiron) [reply src/Logger/Adapter/Raygun.php]: Implemented `getSupportedTypes`, `getSupportedEnvironments`, `getSupportedBreadcrumbTypes` in https://github.com/utopia-php/logger/commit/74140bf8bf9cb8aeb19f7505506e9cd9b71b725b

**Comment by Meldiron** [comment]
@kodumbeats @eldadfux Upgraded requirements to PHP 8.0 in commit https://github.com/utopia-php/logger/commit/db66b8c0c8f8595107cfb093c30b35c170b14887. Please let me know if that is all or I missed something.

File: .travis.yml [hunk 10]
```diff
@@ -0,0 +1,15 @@
+language: php
+
+php:
+- 8.0
+- nightly
```
Comment (christyjacob4) [inline .travis.yml:5]: Do they not have a stable release for php 8.0 yet?

File: src/Logger/Log.php [hunk 11]
```diff
@@ -0,0 +1,338 @@
...
+    }
+
+    /**
+     * Set a type
+     *
+     * @param string $type (required, can be 'log', 'error' or 'warning')
+     * @return void
+     * @throws Exception
+     */
+    public function setType(string $type): void {
```
Comment (christyjacob4) [inline src/Logger/Log.php:95]:
```suggestion
    public function setType(string $type): void 
    {
```

I almost forgot. We need to have the opening brace in a new line according to PHP Fig standards. Please fix it wherever relevant

File: src/Logger/Log.php [hunk 12]
```diff
@@ -0,0 +1,338 @@
...
+
+        $this->type = $type;
+    }
+
+    /**
+     * Get the type of this log
+     *
+     * @return string (can be 'log', 'error' or 'warning')
+     */
+    public function getType(): string {
```
Comment (christyjacob4) [inline src/Logger/Log.php:114]: We need to have the opening brace of a function in a new line according to PHP Fig standards. Please fix it here

File: src/Logger/Log.php [hunk 13]
```diff
@@ -0,0 +1,338 @@
...
+        return $this->type;
+    }
+
+    /**
+     * Set timestamp in seconds when log occurred
+     *
+     * @param float $timestamp (required)
+     * @return void
+     */
+    public function setTimestamp(float $timestamp): void {
```
Comment (christyjacob4) [inline src/Logger/Log.php:124]: We need to have the opening brace of a function in a new line according to PHP Fig standards. Please fix it here

File: src/Logger/Log.php [hunk 14]
```diff
@@ -0,0 +1,338 @@
...
+        $this->timestamp = $timestamp;
+    }
+
+    /**
+     * Get timestamp in seconds when log occurred
+     *
+     * @return float
+     */
+    public function getTimestamp(): float {
```
Comment (christyjacob4) [inline src/Logger/Log.php:133]: We need to have the opening brace of a function in a new line according to PHP Fig standards. Please fix it here

**Review by christyjacob4 — CHANGES_REQUESTED** [review]
Just a minor style comment. 
Please make sure all the opening braces in all the functions start in new lines. Rest looks great! Amazing job 🥳

File: src/Logger/Log.php (hunk 11 above)
Comment (Meldiron) [reply src/Logger/Log.php:95]: Ill fix all manually, but can we make this part of the code-quality test? 🤔 Would be cool prevention for future.

Comment (Meldiron) [reply src/Logger/Log.php:95]: Only class and function blocks, or same for if/for/anything?

Comment (Meldiron) [reply src/Logger/Log.php]: ^ Only func+class

**Comment by eldadfux** [comment]
> @kodumbeats @eldadfux Upgraded requirements to PHP 8.0 in commit [db66b8c](https://github.com/utopia-php/logger/commit/db66b8c0c8f8595107cfb093c30b35c170b14887). Please let me know if that is all or I missed something.

You should be good 👍

File: README.md [hunk 15]
```diff
@@ -1,3 +1,125 @@
...
+$log->addTag('sdk', 'Flutter');
+$log->addTag('sdkVersion', '0.0.1');
+$log->addTag('authMode', 'default');
+$log->addTag('authMethod', 'cookie');
+$log->addTag('authProvider', 'MagicLink');
+$log->addExtra('urgent', false);
+$log->addExtra('isExpected', true);
+
+// Sentry
+$adapter = new Sentry("[YOUR_SENTRY_KEY]", \getenv("[YOUR_SENTRY_PROJECT_ID]"));
```
Comment (eldadfux) [inline README.md:56]:
```suggestion
$adapter = new Sentry(\getenv("[YOUR_SENTRY_KEY]"), \getenv("[YOUR_SENTRY_PROJECT_ID]"));
```

File: src/Logger/Adapter.php [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+
+
+    /**
+     * Push log to external provider
+     *
+     * @param Log $log
+     * @return int
+     * @throws Exception
+     */
+    abstract public function pushLog(Log $log): int;
```
Comment (eldadfux) [inline src/Logger/Adapter.php:24]: I think it will make more sense to call `pushLog` just `push`, taking into account we're already in the logger namespace.

File: src/Logger/Adapter/Raygun.php [hunk 17]
```diff
@@ -0,0 +1,156 @@
+<?php
+
+namespace Utopia\Logger\Adapter;
+
+use Exception;
+use Utopia\Logger\Adapter;
+use Utopia\Logger\Log;
+use Utopia\Logger\Logger;
+use function curl_error;
```
Comment (eldadfux) [inline src/Logger/Adapter/Raygun.php:9]: We usually just use `\` in the function call.

File: src/Logger/Adapter/Raygun.php [hunk 18]
```diff
@@ -0,0 +1,156 @@
...
+use function curl_getinfo;
+use const CURLINFO_HTTP_CODE;
```
Comment (eldadfux) [inline src/Logger/Adapter/Raygun.php:11]: Same for const, we just call it.

File: README.md [hunk 19]
```diff
@@ -1,3 +1,125 @@
...
+$log->addBreadcrumb(new Breadcrumb(Log::TYPE_INFO, "auth", "Authenticated with * Using API Key", \microtime(true) - 350));
+$log->addBreadcrumb(new Breadcrumb(Log::TYPE_INFO, "database", "Found collection abcd1234", \microtime(true) - 300));
+$log->addBreadcrumb(new Breadcrumb(Log::TYPE_DEBUG, "database", "Permission for collection abcd1234 met", \microtime(true) - 200));
+$log->addBreadcrumb(new Breadcrumb(Log::TYPE_ERROR, "database", "Missing document when searching by ID!", \microtime(true)));
+$log->addTag('sdk', 'Flutter');
+$log->addTag('sdkVersion', '0.0.1');
+$log->addTag('authMode', 'default');
+$log->addTag('authMethod', 'cookie');
+$log->addTag('authProvider', 'MagicLink');
+$log->addExtra('urgent', false);
```
Comment (eldadfux) [inline README.md:52]: Can we think of a better name than Extra? Seems very generic and not intuitive. Why can't we use `addTag()` for this data as well?

**Review by eldadfux — COMMENTED** [review]
This is beautiful work. The codebase looks much cleaner with the new changes, and implementation of a new adapter looks much easier. I added a few comments, most of them regarding naming changes.

File: README.md (hunk 19 above)
Comment (christyjacob4) [reply README.md:52]: That's probably only because `tags` and `extras` are displayed differently in the UIs of these adapters.

File: src/Logger/Adapter/Raygun.php (hunk 17 above)
Comment (Meldiron) [reply src/Logger/Adapter/Raygun.php:9]: My bad, PHPStorm likes it this way 😅 Updated.

File: src/Logger/Adapter/Raygun.php (hunk 18 above)
Comment (Meldiron) [reply src/Logger/Adapter/Raygun.php:11]: Updated

File: README.md (hunk 19 above)
Comment (Meldiron) [reply README.md:52]: As @christyjacob4 mentioned, some APIs supports them and it makes sense to me. Tags are like a label so you can easily filter, sort, search... Extra is more like an JSON object with bunch of additional data.

For example, `version` is great example of `tag`, because you want to filter by this information.
`contactFormSubject` is great example of `extra`, because it might help you debug the error, but you will never search by **specific** text user wrote.

In other words, `tags` is there for finding a specific error, `extra` is there for understanding the error.

Comment (eldadfux) [reply README.md:52]: Functionality wise, I'm happy with it, just the name feels not 100%. We can move on with this and re-iterate when we have a better idea.

**Review by eldadfux — APPROVED** [review]
